validator/coordinator/src/test/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinatorTest.java

```diff
@@ -13,54 +13,115 @@
+  private AttestationAggregator attestationAggregator;
+  private EventBus eventBus;
+  private ChainStorageClient storageClient;
+  private ArtemisConfiguration config;
   private ValidatorCoordinator vc;
 
-  @Test
-  void noValidators_noAttestationProduced() throws Exception {
-    int numValidators = 0;
+  private final int numValidators = 12;
 
-    ArtemisConfiguration config = mock(ArtemisConfiguration.class);
+  @BeforeEach
+  void setup() {
+    config = mock(ArtemisConfiguration.class);
     doReturn(0).when(config).getNaughtinessPercentage();
     doReturn(numValidators).when(config).getNumValidators();
     doReturn(null).when(config).getValidatorsKeyFile();
     doReturn(0).when(config).getInteropOwnedValidatorStartIndex();
     doReturn(numValidators).when(config).getInteropOwnedValidatorCount();
 
-    EventBus eventBus = spy(new EventBus());
-    AttestationAggregator attestationAggregator = mock(AttestationAggregator.class);
-    BlockAttestationsPool blockAttestationsPool = mock(BlockAttestationsPool.class);
-    final ChainStorageClient storageClient = new ChainStorageClient(eventBus);
-    final BeaconChainUtil chainUtil = BeaconChainUtil.create(numValidators, storageClient);
+    attestationAggregator = mock(AttestationAggregator.class);
+    blockAttestationsPool = mock(BlockAttestationsPool.class);
+    doReturn(createAttestations())
+        .when(blockAttestationsPool)
+        .getAggregatedAttestationsForBlockAtSlot(any());
+
+    eventBus = spy(new EventBus());
+    storageClient = new ChainStorageClient(eventBus);
+    List<BLSKeyPair> blsKeyPairList =
+        new MockStartValidatorKeyPairFactory().generateKeyPairs(0, numValidators);
+    final BeaconChainUtil chainUtil = BeaconChainUtil.create(storageClient, blsKeyPairList);
     chainUtil.initializeStorage();
 
     vc =
         spy(
             new ValidatorCoordinator(
                 eventBus, storageClient, attestationAggregator, blockAttestationsPool, config));
+  }
 
+  @Test
+  void onAttestationEvent_noAttestationAssignments() throws Exception {
+    doReturn(0).when(config).getInteropOwnedValidatorCount();
+    vc =
+        spy(
+            new ValidatorCoordinator(
+                eventBus, storageClient, attestationAggregator, blockAttestationsPool, config));
     eventBus.post(new BroadcastAttestationEvent(storageClient.getBestBlockRoot()));
 
     // Until the PR #1043 gets merged in that contains "ensureConditionRemainsMet"
     Thread.sleep(1000);
     verify(attestationAggregator, never()).updateAggregatorInformations(any());
     verify(vc, never()).asyncProduceAttestations(any(), any(), any());
   }
+
+  @Test
+  void createBlockAfterNormalSlot() {
+    eventBus.post(new SlotEvent(storageClient.getBestSlot().plus(UnsignedLong.ONE)));
+    ArgumentCaptor<BeaconBlock> blockArgumentCaptor = ArgumentCaptor.forClass(BeaconBlock.class);
+    Waiter.waitFor(() -> verify(eventBus, atLeastOnce()).post(blockArgumentCaptor.capture()));
```

Comment:
This is unfortunately brittle because `ArgumentCaptor` doesn't check the argument type.  The test will fail if we ever start firing a different type of event first.

Also thankfully block creation is done on the calling thread so we don't need the Waiter.waitFor here.  So you can just use:
```
  @Test
  void createBlockAfterSkippedSlot() {
    final UnsignedLong slotNumber = UnsignedLong.valueOf(2);
    eventBus.post(new SlotEvent(storageClient.getBestSlot().plus(slotNumber)));
    verify(eventBus, atLeastOnce()).post(blockWithSlot(slotNumber));
  }

  private Object blockWithSlot(final UnsignedLong slotNumber) {
    return argThat(
        argument -> {
          if (!(argument instanceof BeaconBlock)) {
            return false;
          }
          final BeaconBlock block = (BeaconBlock) argument;
          return block.getSlot().equals(slotNumber);
        });
  }
```

and then reuse `blockWithSlot` in the two tests below as well.

Reply:
Actually, we do fire 4 different events on the same eventBus before firing the block, and argument captor returns the actual block instead of typecasting a different event to a block. I will still do as you suggest above, but it's interesting that the argument captor actually does seem to act as if its type aware in this instance. Even if it's not supposed to be?

Reply:
Do you mind explaining a bit more why we don't need to use the Waiter.waitFor ? I'm a bit confused.

Reply:
Done. but would appreciate if you could still respond to my questions.

Reply:
The argument captor `getValue()` method returns the last value it caught so if the block is fired last it happens to work (I thought it was the other way around originally).

We don't need the Waiter.waitFor because we're using a plain `EventBus` not an `AsyncEventBus` so the thread that posts the `SlotEvent` is the same thread used to call `ValidatorCoordinator.onNewSlot` which you can follow through to see it stays on that same thread until it posts the new block event.  So the call to `post(new SlotEvent(...))` won't return until after the new block has been created and posted.

The waiter is required for `BroadcastAttestationEvent` because it winds up calling `asyncProduceAttestations` which jumps to different threads to produce the attestations.

Reply:
thank you for the explanation.